pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-tenant/multiregion-9node-3region-3azs-tenant_test: TestCCLLogic_regional_by_row_foreign_key failed

pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-tenant/multiregion-9node-3region-3azs-tenant_test.TestCCLLogic_regional_by_row_foreign_key [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21001031?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21001031?buildTab=artifacts#/) on release-26.1.0-rc @ [6c9b2f778bd29895383d493a2fa4490323499ecf](https://github.com/cockroachdb/cockroach/commits/6c9b2f778bd29895383d493a2fa4490323499ecf):


```
=== RUN   TestCCLLogic_regional_by_row_foreign_key
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/9010d58e1c7b3e4e95937947c4203bde/logTestCCLLogic_regional_by_row_foreign_key4268473012
    test_log_scope.go:82: use -show-logs to present logs inline
[19:31:23] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3270/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-tenant/multiregion-9node-3region-3azs-tenant_test_/multiregion-9node-3region-3azs-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key with config multiregion-9node-3region-3azs-tenant: 9 tests, 0 failures
    logic.go:4583: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3270/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-tenant/multiregion-9node-3region-3azs-tenant_test_/multiregion-9node-3region-3azs-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key:64: error while processing
    logic.go:4583: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3270/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-tenant/multiregion-9node-3region-3azs-tenant_test_/multiregion-9node-3region-3azs-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_foreign_key:64: 
        expected success, but found
        (XX000) internal error: comparison of two different versions of enum USER DEFINED ENUM: crdb_internal_region oid 100108: versions 6 and 7
        datum.go:5313: in Compare()
        DETAIL: stack trace:
        pkg/sql/sem/tree/datum.go:5313: Compare()
        pkg/sql/sem/eval/binary_op.go:158: EvalCompareScalarOp()
        bazel-out/s390x-fastbuild/bin/pkg/sql/sem/tree/eval_op_generated.go:307: Eval()
        pkg/sql/sem/eval/expr.go:280: EvalComparisonExpr()
        bazel-out/s390x-fastbuild/bin/pkg/sql/sem/tree/eval_expr_generated.go:116: Eval()
        pkg/sql/sem/eval/expr.go:570: EvalOrExpr()
        bazel-out/s390x-fastbuild/bin/pkg/sql/sem/tree/eval_expr_generated.go:341: Eval()
        pkg/sql/sem/eval/expr.go:22: Expr()
        pkg/sql/execinfra/execexpr/expr.go:44: RunFilter()
        pkg/sql/execinfra/execexpr/expr.go:293: evalFilter()
        pkg/sql/execinfra/execexpr/expr.go:97: EvalFilter()
        pkg/sql/rowexec/joinerbase.go:232: render()
        pkg/sql/rowexec/joinreader_strategies.go:289: nextRowToEmit()
        pkg/sql/rowexec/joinreader.go:1237: emitRow()
        pkg/sql/rowexec/joinreader.go:815: Next()
        pkg/sql/colexec/columnarizer.go:248: Next()
        pkg/sql/colexec/invariants_checker.go:88: Next()
        pkg/sql/colflow/stats.go:120: next()
        pkg/sql/colexecerror/error.go:162: CatchVectorizedRuntimeError()
        pkg/sql/colflow/stats.go:128: Next()
        pkg/sql/colexec/materializer.go:246: next()
        pkg/sql/colexec/materializer.go:271: nextAdapter()
        pkg/sql/colexecerror/error.go:162: CatchVectorizedRuntimeError()
        pkg/sql/colexec/materializer.go:277: Next()
        pkg/sql/row_source_to_plan_node.go:80: Next()
        pkg/sql/error_if_rows.go:37: Next()
        pkg/sql/plan_node_to_row_source.go:223: Next()
        pkg/sql/colexec/columnarizer.go:248: Next()
        pkg/sql/colexec/invariants_checker.go:88: Next()
        pkg/sql/colflow/stats.go:120: next()
        pkg/sql/colexecerror/error.go:162: CatchVectorizedRuntimeError()
        pkg/sql/colflow/stats.go:128: Next()
        
        NOTE: internal errors may have more details in logs. Use -show-logs.
        
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9010d58e1c7b3e4e95937947c4203bde/logTestCCLLogic_regional_by_row_foreign_key4268473012
--- FAIL: TestCCLLogic_regional_by_row_foreign_key (7.17s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCCLLogic_regional_by_row_foreign_key.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58925

--- a/pkg/sql/catalog/descs/collection.go
+++ b/pkg/sql/catalog/descs/collection.go
@@ -151,6 +151,17 @@ type Collection struct {
 	// txn. This guarantees the generation for long-running transactions
 	// this value stays the same for the life of the transaction.
 	leaseGeneration int64
+
+	// forceStorageLookupIDs contains descriptor IDs that must bypass all
+	// non-storage layers (leased, cached, uncommitted, synthetic) and be
+	// resolved directly from KV. This is used when a fresh Collection is
+	// created for parallel check worker goroutines: the parent (planner's)
+	// Collection has uncommitted descriptors that are only visible via KV
+	// within the same transaction. Without this, a leased or cached descriptor
+	// with a stale version could be returned.
+	// When set, the collection is implicitly read-only to avoid multiple sources
+	// of uncommitted descriptor state.
+	forceStorageLookupIDs catalog.DescriptorIDSet
 }
 
 // FromTxn is a convenience function to extract a descs.Collection which is
@@ -332,6 +343,16 @@ func (tc *Collection) CountUncommittedNewOrDroppedDescriptors() int {
 		return nil
 	})
 	return count
+}
+
+// GetUncommittedDescriptorIDs returns the IDs of all uncommitted descriptors.
+func (tc *Collection) GetUncommittedDescriptorIDs() catalog.DescriptorIDSet {
+	var ids catalog.DescriptorIDSet
+	_ = tc.uncommitted.iterateUncommittedByID(func(desc catalog.Descriptor) error {
+		ids.Add(desc.GetID())
+		return nil
+	})
+	return ids
 }
 
 // HasUncommittedTypes returns true if the Collection contains uncommitted
@@ -562,6 +583,11 @@ func (tc *Collection) WriteDescToBatch(
 	b *kv.Batch,
 	opts ...WriteDescOption,
 ) error {
+	if !tc.forceStorageLookupIDs.Empty() {
+		return errors.AssertionFailedf("descriptor collection with " +
+			"forceStorageLookupIDs set is read-only, should not be used to write " +
+			"descriptors")
+	}
 	if desc.GetID() == descpb.InvalidID {
 		return errors.AssertionFailedf("cannot write descriptor with an empty ID: %v", desc)
 	}

--- a/pkg/sql/catalog/descs/collection_test.go
+++ b/pkg/sql/catalog/descs/collection_test.go
@@ -1296,3 +1296,106 @@ func getDatabaseNames(allDBs []catalog.DatabaseDescriptor) []string {
 	}
 	return names
 }
+
+// TestForceStorageLookupIDs verifies that when a Collection is created with
+// WithForceStorageLookupIDs, descriptor lookups for IDs in the forced set
+// bypass the leased/cached layers and resolve directly from KV storage.
+func TestForceStorageLookupIDs(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	srv, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer srv.Stopper().Stop(ctx)
+	s := srv.ApplicationLayer()
+
+	tdb := sqlutils.MakeSQLRunner(sqlDB)
+	tdb.Exec(t, "CREATE DATABASE db")
+	tdb.Exec(t, "USE db")
+	// Here the descriptor of db.public.typ is commited with initial version = 1.
+	tdb.Exec(t, "CREATE TYPE db.public.typ AS ENUM ('a')")
+
+	execCfg := s.ExecutorConfig().(sql.ExecutorConfig)
+
+	require.NoError(t, sql.DescsTxn(ctx, &execCfg, func(
+		ctx context.Context, txn isql.Txn, parentCollection *descs.Collection,
+	) error {
+		tn := tree.MakeQualifiedTypeName("db", "public", "typ")
+		_, typ, err := descs.PrefixAndMutableType(ctx, parentCollection.MutableByName(txn.KV()), &tn)
+		require.NoError(t, err)
+		committedVersion := typ.Version
+
+		// Here `WriteDesc()` calls `MaybeIncrementVersion()` that increase the UDT
+		// version by 1.
+		// It also writes to KV within this transaction but without committing.
+		require.NoError(t, parentCollection.WriteDesc(ctx, false /* kvTrace */, typ, txn.KV()))
+		kvVersion := typ.Version
+		require.Greater(t, kvVersion, committedVersion,
+			"WriteDesc should have incremented the version")
+		typID := typ.GetID()
+
+		// Create the child collection with forceStorageLookupIDs set to the
+		// parent's uncommitted descriptor IDs.
+		uncommittedIDs := parentCollection.GetUncommittedDescriptorIDs()
+		require.True(t, uncommittedIDs.Contains(typID),
+			"type descriptor should be in uncommitted IDs")
+
+		childCollection := execCfg.CollectionFactory.NewCollection(
+			ctx, descs.WithForceStorageLookupIDs(uncommittedIDs),
+		)
+		defer childCollection.ReleaseAll(ctx)
+
+		// Also create a child without forceStorageLookupIDs to contrast.
+		childCollectionNoForce := execCfg.CollectionFactory.NewCollection(ctx)
+		defer childCollectionNoForce.ReleaseAll(ctx)
+
+		nonExistentID := descpb.ID(999999)
+
+		testCases := []struct {
+			name            string
+			collection      *descs.Collection
+			id              descpb.ID
+			expectedVersion descpb.DescriptorVersion // 0 means don't check (error case)
+			expectErrStr    string
+		}{
+			{
+				name:            "forced ID resolves from KV storage",
+				collection:      childCollection,
+				id:              typID,
+				expectedVersion: kvVersion,
+			},
+			{
+				name:            "non-forced child gets stale leased version",
+				collection:      childCollectionNoForce,
+				id:              typID,
+				expectedVersion: committedVersion,
+			},
+			{
+				name:         "forced child returns error for non-existent ID",
+				collection:   childCollection,
+				id:           nonExistentID,
+				expectErrStr: "descriptor not found",
+			},
+			{
+				name:         "non-forced child returns error for non-existent ID",
+				collection:   childCollectionNoForce,
+				id:           nonExistentID,
+				expectErrStr: "descriptor not found",
+			},
+		}
+
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				desc, err := tc.collection.ByIDWithLeased(txn.KV()).WithoutNonPublic().Get().Desc(ctx, tc.id)
+				if tc.expectErrStr != "" {
+					require.Contains(t, err.Error(), tc.expectErrStr)
+					return
+				}
+				require.NoError(t, err)
+				require.Equal(t, tc.expectedVersion, desc.GetVersion())
+			})
+		}
+
+		return nil
+	}))
+}

--- a/pkg/sql/catalog/descs/descriptor.go
+++ b/pkg/sql/catalog/descs/descriptor.go
@@ -483,6 +483,9 @@ func (q *byIDLookupContext) lookupLeased(
 	if q.flags.layerFilters.withoutLeased || lease.TestingTableLeasesAreDisabled() {
 		return nil, catalog.NoValidation, nil
 	}
+	if q.tc.forceStorageLookupIDs.Contains(id) {
+		return nil, catalog.NoValidation, nil
+	}
 	// If we have already read all of the descriptors, use it as a negative
 	// cache to short-circuit a lookup we know will be doomed to fail.
 	if q.tc.cr.IsDescIDKnownToNotExist(id, q.flags.descFilters.maybeParentID) {

--- a/pkg/sql/catalog/descs/factory.go
+++ b/pkg/sql/catalog/descs/factory.go
@@ -12,6 +12,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/spanconfig"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/hydrateddesccache"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv"
@@ -103,8 +104,9 @@ func NewBareBonesCollectionFactory(
 }
 
 type constructorConfig struct {
-	dsdp    DescriptorSessionDataProvider
-	monitor *mon.BytesMonitor
+	dsdp                  DescriptorSessionDataProvider
+	monitor               *mon.BytesMonitor
+	forceStorageLookupIDs catalog.DescriptorIDSet
 }
 
 // Option is how optional construction parameters are provided to the
@@ -126,6 +128,17 @@ func WithDescriptorSessionDataProvider(
 func WithMonitor(monitor *mon.BytesMonitor) func(b *constructorConfig) {
 	return func(cfg *constructorConfig) {
 		cfg.monitor = monitor
+	}
+}
+
+// WithForceStorageLookupIDs configures the Collection to bypass all
+// non-storage layers (leased, cached, etc.) for the given descriptor IDs,
+// resolving them directly from KV. This is used when a fresh Collection is
+// created for parallel check workers whose parent Collection has uncommitted
+// descriptors visible only via KV within the same transaction.
+func WithForceStorageLookupIDs(ids catalog.DescriptorIDSet) func(cfg *constructorConfig) {
+	return func(cfg *constructorConfig) {
+		cfg.forceStorageLookupIDs = ids
 	}
 }
 
@@ -162,6 +175,7 @@ func (cf *CollectionFactory) NewCollection(ctx context.Context, options ...Optio
 		cr:                      catkv.NewCatalogReader(cf.codec, v, cf.systemDatabase, cfg.monitor),
 		temporarySchemaProvider: cfg.dsdp,
 		validationModeProvider:  cfg.dsdp,
+		forceStorageLookupIDs:   cfg.forceStorageLookupIDs,
 	}
 }
 

--- a/pkg/sql/distsql/server.go
+++ b/pkg/sql/distsql/server.go
@@ -545,9 +545,13 @@ func (ds *ServerImpl) newFlowContext(
 		// responsible for cleaning it up and releasing any accessed descriptors
 		// on flow cleanup.
 		dsdp := catsessiondata.NewDescriptorSessionDataStackProvider(evalCtx.SessionDataStack)
-		flowCtx.Descriptors = ds.CollectionFactory.NewCollection(
-			ctx, descs.WithDescriptorSessionDataProvider(dsdp),
-		)
+		opts := []descs.Option{descs.WithDescriptorSessionDataProvider(dsdp)}
+		if localState.Collection != nil {
+			opts = append(opts, descs.WithForceStorageLookupIDs(
+				localState.Collection.GetUncommittedDescriptorIDs(),
+			))
+		}
+		flowCtx.Descriptors = ds.CollectionFactory.NewCollection(ctx, opts...)
 		flowCtx.IsDescriptorsCleanupRequired = true
 		var evalCatalogBuiltins evalcatalog.Builtins
 		// In distributed execution, authorization was already checked on the gateway node.

